src/kOS.Safe/Encapsulation/ScalarValue.cs

```diff
@@ -0,0 +1,456 @@
+﻿using System;
+using kOS.Safe.Encapsulation.Suffixes;
+using kOS.Safe.Exceptions;
+
+namespace kOS.Safe.Encapsulation
+{
+    abstract public class ScalarValue : Structure, IConvertible
+    {
+        protected object internalValue;
+
+        abstract public bool IsInt { get; }
+
+        abstract public bool IsDouble { get; }
+
+        public bool IsValid
+        {
+            get
+            {
+                if (IsInt) return true;
+                if (IsDouble)
+                {
+                    double d = GetDoubleValue();
+                    if (double.IsInfinity(d) || double.IsNaN(d)) return false;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public object Value { get { return internalValue; } }
+
+        protected ScalarValue()
+        {
+            InitializeSuffixes();
+        }
+
+        public void InitializeSuffixes()
+        {
+            AddSuffix("ISINT", new Suffix<bool>(() => IsInt));
```

Comment:
Suffixes were removed in the combined PR instead of this one, mostly because I didn't want to do another round of branch switching.  I can make that change in this PR itself if it will make it easier for you.